src/main/java/bio/terra/service/snapshotbuilder/SnapshotBuilderService.java

```diff
@@ -206,51 +208,75 @@ private EnumerateSnapshotAccessRequest convertToEnumerateModel(
-  }
+  record ParentQueryResult(int parentId, int childId, String childName) {
+    ParentQueryResult(ResultSet rs) throws SQLException {
+      this(
+          rs.getInt(HierarchyQueryBuilder.PARENT_ID),
+          rs.getInt(HierarchyQueryBuilder.CONCEPT_ID),
+          rs.getString(HierarchyQueryBuilder.CONCEPT_NAME));
+    }
 
-  private static SnapshotBuilderConcept getConcept(List<SnapshotBuilderConcept> concepts, int id) {
-    return concepts.stream().filter(c -> c.getId().equals(id)).findFirst().orElseThrow();
+    ParentQueryResult(FieldValueList row) {
+      this(
+          (int) row.get(HierarchyQueryBuilder.PARENT_ID).getLongValue(),
+          (int) row.get(HierarchyQueryBuilder.CONCEPT_ID).getLongValue(),
+          row.get(HierarchyQueryBuilder.CONCEPT_NAME).getStringValue());
+    }
   }
 
-  // Map of concept id to parent concept id.
-  private static final Map<Integer, Integer> HIERARCHY =
-      Map.of(406, 404, 405, 404, 404, 401, 403, 401, 407, 402, 401, 400, 400, 100, 408, 100);
-
   public SnapshotBuilderGetConceptHierarchyResponse getConceptHierarchy(
-      UUID id, int conceptId, AuthenticatedUserRequest userRequest) {
-    var concepts = loadConcepts();
-    SnapshotBuilderConcept concept = getConcept(concepts, conceptId);
-    while (true) {
-      // For a child concept, generate its parent concept and siblings until we reach the root.
-      var parentId = HIERARCHY.get(concept.getId());
-      if (parentId == null) {
-        break;
-      }
-      var parent = getConcept(concepts, parentId);
-      concept =
-          parent.children(
-              HIERARCHY.entrySet().stream()
-                  .filter(e -> e.getValue().equals(parentId))
-                  .map(entry -> getConcept(concepts, entry.getKey()))
-                  .toList());
+      UUID datasetId, int conceptId, AuthenticatedUserRequest userRequest) {
+    Dataset dataset = datasetService.retrieve(datasetId);
+    var query =
+        queryBuilderFactory
+            .hierarchyQueryBuilder(getTableNameGenerator(dataset, userRequest))
+            .generateQuery(conceptId);
+    var sql = query.renderSQL(CloudPlatformWrapper.of(dataset.getCloudPlatform()));
+
+    Map<Integer, SnapshotBuilderParentConcept> parents = new HashMap<>();
+    runSnapshotBuilderQuery(sql, dataset, ParentQueryResult::new, ParentQueryResult::new)
+        .forEach(
+            row -> {
+              SnapshotBuilderParentConcept parent =
+                  parents.computeIfAbsent(
+                      row.parentId,
+                      k ->
+                          new SnapshotBuilderParentConcept()
+                              .parentId(k)
+                              .children(new ArrayList<>()));
+              parent.addChildrenItem(
+                  new SnapshotBuilderConcept()
+                      .id(row.childId)
+                      .name(row.childName)
+                      .hasChildren(true)
+                      .count(1));
```

Comment:
is there a ticket for count too? I don't see one in the backlog

Reply:
That's a good point, I created a ticket https://broadworkbench.atlassian.net/browse/DC-951